Fix USB_UART_CONSOLE option with shell and logs

aurel32 opened the pull request:
This patch series fixes the USB_UART_CONSOLE option providing _USB port for console outputs_ so that it can be used for shell and logs. It has been tested on NRF52840 dongle and SAM E70 Xplained. It is especially interesting on the NRF52840 dongle as there is no easy access to the serial port.

aurel32 committed:
drivers: console: fix USB_UART_CONSOLE

The USB_UART_CONSOLE option itself is not enough to enable console on
the USB UART port. Most of the code is actually shared with
UART_CONSOLE, so this symbol has to be selected.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: cdc_acm: fix interrupt mode

The CDC ACM driver has to emulate the TX and RX interrupts from the USB
ones. However it does not correctly emulate them as "TX buffer empty"
and "RX buffer not empty" interrupts.

For "TX buffer empty" interrupt:
- Reading the interrupt status should not clear the interrupt
- Enabling the interrupt should fire an interrupt if the TX buffer is
  empty
- An interrupt should be triggered when the USB device get configured

For "RX buffer not empty" interrupt:
- Reading the interrupt status should not clear the interrupt
- Enabling the interrupt should fire an interrupt if the RX buffer is
  not empty

This make the a console on the USB CDC ACM port usable for the shell
and the logs.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

aurel32 committed:
usb: cdc_acm: select SERIAL_HAS_DRIVER and SERIAL_SUPPORT_INTERRUPT

USB CDC ACM is a serial driver like another. Therefore select both
SERIAL_HAS_DRIVER and SERIAL_SUPPORT_INTERRUPT. This allows one to
enable the console driver without having to enable another serial
driver.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

finikorg reviewed (approved):
LGTM

aurel32 commented on samples/subsys/usb/console/nrf52840_pca10059.overlay:
```diff
@@ -0,0 +1,10 @@
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Well if the following snipped is not added to the device tree of the soc/board, the sample will not build:

```
                usb_cdc: virtualcom {
                        compatible = "nordic,nrf-usbd";
                        label = "CDC_ACM";
                };
```

On the other hand the sample will build but won't work without this overlay.

Alternatively I don't know if we could easily use symlink so that at least the same file is used.

finikorg replied:
We have some boards with default console over USB (tinytile), now we need to put those overlays to every sample?

finikorg replied:
Before it was in Kconfig, everything was much better

finikorg replied:
But this does not affect this PR

aurel32 replied:
> But this does not affect this PR

I didn't feel like separating this last commit from the others, as it's basically how I tested it. Now if it requires some discussion, I can just drop it from this PR and move it to a new one.

finikorg replied:
Leave it here, we might fix this later.

jfischer-phytec-iot replied:
The same overlay can also be used for usb-kw24d512 board. Just wonder what virtualcom is and where it comes from...

aurel32 replied:
Ok. As changing that seems controversial, I have just dropped the nrf52840 dongle related commits. I'll do a separate pull request once this PR is merged.

pfalcon commented:
Btw guys, I don't know how we ended up with "USB console". There should be just one console - subsys/console. Anything else would need to go (unless there're sisyphuses to roll similar but different things on and on).

aurel32 commented:
> Btw guys, I don't know how we ended up with "USB console". There should be just one console - subsys/console. Anything else would need to go (unless there're sisyphuses to roll similar but different things on and on).

Maybe it's a misnomer. By USB console, I mean setting up `drivers/console` so that it uses a USB CDC ACM UART instead of a more standard UART.

pfalcon commented:
> Maybe it's a misnomer. By USB console, I mean setting up drivers/console

Yep, we have too many console in Zephyr indeed. I see drivers/console to go too - but not before subsys/console is fully ready to replace it, which is not too soon...  Thanks for reply.

aurel32 commented:
I have just added an additional commit changing the defconfig of the nRF58240 dongle to remove some entries that are not necessary anymore.

aurel32 commented:
I have just dropped the controversial nrf52840 dongle related commits, and updated the title and the description of the PR to avoid using _USB console_.

pfalcon commented:
> and updated the title and the description of the PR to avoid using USB console.

Much clearer that way, thanks. I wonder if it works with samples/subsys/console/echo . I'm not sure how helpful it is, but let me approve it.

aurel32 commented:
> I wonder if it works with samples/subsys/console/echo .

After adding the DTS overlay and tweaking `prj.conf` as follow, this sample works.

```
CONFIG_USB=y
CONFIG_USB_DEVICE_STACK=y
CONFIG_USB_DEVICE_PRODUCT="Zephyr USB console sample"
CONFIG_USB_UART_CONSOLE=y
CONFIG_UART_INTERRUPT_DRIVEN=y
CONFIG_UART_LINE_CTRL=y
CONFIG_USB_UART_DTR_WAIT=y
```

pfalcon commented:
Cool, thanks for confirming!

nashif commented:
need rebase and there is a typo in the commit message.

aurel32 commented:
@nashif: done
